grpc/src/tx.rs

```diff
@@ -0,0 +1,602 @@
+    pub async fn with_url(
+        url: impl Into<String>,
+        account_address: &Address,
+        account_pubkey: VerifyingKey,
+        signer: S,
+    ) -> Result<Self> {
+        let transport = tonic::transport::Endpoint::from_shared(url.into())
+            .map_err(|e| Error::TransportError(e.to_string()))?
+            .connect_lazy();
+        Self::new(transport, account_address, account_pubkey, signer).await
+    }
+}
+
+#[cfg(target_arch = "wasm32")]
+impl<S> TxClient<tonic_web_wasm_client::Client, S>
+where
+    S: Signer<Signature>,
+{
+    /// Create a new client connected to the given `url` with default
+    /// settings of [`tonic_web_wasm_client::Client`].
+    pub async fn with_grpcweb_url(
+        url: impl Into<String>,
+        account_address: &Address,
+        account_pubkey: VerifyingKey,
+        signer: S,
+    ) -> Result<Self> {
+        let transport = tonic_web_wasm_client::Client::new(url.into());
+        Self::new(transport, account_address, account_pubkey, signer).await
+    }
+}
+
+impl<T, S> Deref for TxClient<T, S> {
+    type Target = GrpcClient<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+impl<T, S> fmt::Debug for TxClient<T, S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("TxClient { .. }")
+    }
+}
+
+/// Sign `tx_body` and the transaction metadata as the `base_account` using `signer`
+pub fn sign_tx(
+    tx_body: RawTxBody,
+    chain_id: Id,
+    base_account: &BaseAccount,
+    verifying_key: &VerifyingKey,
+    signer: &impl Signer<Signature>,
+    gas_limit: u64,
+    fee: u64,
+) -> RawTx {
+    // From https://github.com/celestiaorg/cosmos-sdk/blob/v1.25.0-sdk-v0.46.16/proto/cosmos/tx/signing/v1beta1/signing.proto#L24
+    const SIGNING_MODE_INFO: ModeInfo = ModeInfo {
+        sum: Sum::Single { mode: 1 },
+    };
+
+    let public_key = secp256k1::PubKey {
+        key: verifying_key.to_encoded_point(true).as_bytes().to_vec(),
+    };
+    let public_key_as_any = Any {
+        type_url: secp256k1::PubKey::type_url(),
+        value: public_key.encode_to_vec(),
+    };
+
+    let mut fee = Fee::new(fee, gas_limit);
+    fee.payer = Some(base_account.address.clone());
+
+    let auth_info = AuthInfo {
+        signer_infos: vec![SignerInfo {
+            public_key: Some(public_key_as_any),
+            mode_info: SIGNING_MODE_INFO,
+            sequence: base_account.sequence,
+        }],
+        fee,
+    };
+
+    let bytes_to_sign = SignDoc {
+        body_bytes: tx_body.encode_to_vec(),
+        auth_info_bytes: auth_info.clone().encode_vec(),
+        chain_id: chain_id.into(),
+        account_number: base_account.account_number,
+    }
+    .encode_to_vec();
+
+    let signature = signer.sign(&bytes_to_sign);
+
+    RawTx {
+        auth_info: Some(auth_info.into()),
+        body: Some(tx_body),
+        signatures: vec![signature.to_bytes().to_vec()],
+    }
+}
```

Comment:
moved from https://github.com/eigerco/lumina/pull/474/files#diff-62aed61373fd1face63620b93a60741695b8fa76cbc964fd013d7b06bd656a08L83-L130
with changes to arguments types